### PR TITLE
(PC-16114) docs(testStrategy): initiate new strategy

### PR DESCRIPTION
**Pull request by marineshaw**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-16114

### CODE REVIEW

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+- Timely : Test should be written before their feature (TDD)
+
+---
+
+## When to use .web and .native?
+
+use `MyComponent.native.test.tsx` if
+
+- the logic is shared between web and native
+- it exists a file `MyComponent.native.tsx.`
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:19]:
```suggestion
- there is a file `MyComponent.native.tsx.`
```

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+use `MyComponent.web.test.tsx` if
+
+- it exists a file `MyComponent.web.tsx`
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:23]:
```suggestion
- there is a file `MyComponent.web.tsx`
```

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+- the logic and/or design differs from web to native
+
+  - if you used `Platform.OS === web`, `isWeb`
+  - if you used `isTouch` or `isNative`
+    -> in that case, you should use userEvent instead of fireEvent
+
+use `MyComponent.test.ts` if there is no render, for example when testing a hook
+
+**Do not use** `MyComponent.test.tsx` without `.web` or `.native`
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:32]: Ces 2 phrases portent un peu à confusion, la première dit d'utiliser `MyComponent.test.ts` si c'est un hook et la seconde dit de ne surtout pas utiliser `MyComponent.test.tsx`. On pourrait renommer le premier en `useSomething.test.ts` ou un truc comme ça ?

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+**Storybook** is used when testing
+
+- a component that is used multiple times in the app
+- a component from the design system
+- a component which size differs depending on the device
+
+**Storybook** is not used when testing
+
+- a big and complex component
+- a component used in a single page
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:57]: Ah ? Je pensais qu'à terme storybook servirait de cheatcodes pour le web un peu, et qu'on pourrait y mettre tous nos composants, même ceux dédiés à une seule page. Par exemple on a commencé à mettre les composants de la home dedans

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+---
+
+## When to use queryBy, findBy and getBy
+
+- Prove that an element exists with `getBy` / `queryBy` + `expect(…).toBeTruthy()`
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:63]: Pourquoi ne pas recommander seulement `getBy` ici ? D'après la doc sur queryBy : `This is useful for asserting an element that is not present.`

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+- Prove that an element does not exist with `queryBy` + `expect(…).toBeNull()`
+- Query an element to process some action on it : `getBy` + action
+- **Do not** use `waitFor(getBy)`or `waitFor(queryBy)`, use `findBy`instead
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:66]: Pourquoi ? Si `findBy` = `waitFor(getBy)` ça ne me semble pas très grave d'utiliser l'un ou l'autre si ? On peut formuler ça comme une recommandation, genre "`findBy` is equivalent to `waitFor(getBy)`"

File: doc/standards/testStrategy/TestCookBook.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+#### Authentication with `authContext`
+
+To test as an authenticated person, you can mock `useAuthContext` as follows:
+
+```tsx
+import { useAuthContext } from 'features/auth/AuthContext'
+jest.mock('features/auth/AuthContext')
+const mockUseAuthContext = useAuthContext as jest.MockedFunction<typeof useAuthContext>
+// ... and then within each test
```
Comment (anoukhello) [inline doc/standards/testStrategy/TestCookBook.md:27]: Cette phrase me donne envie de mettre un `mockReturnValue**Once**` juste en dessous

File: doc/standards/testStrategy/TestCookBook.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  setIsLoggedIn: jest.fn(),
+  user: anUserFixtures,
+  refetchUser: jest.fn(),
+  isUserLoading: false,
+})
+```
+
+#### Mock route params
+
+When the tested component use route params through `useRoute` hook, `params` can be mocked like follow:
```
Comment (anoukhello) [inline doc/standards/testStrategy/TestCookBook.md:39]:
```suggestion
When the tested component use route params through `useRoute` hook, `params` can be mocked as follows:
```

**Review by anoukhello — COMMENTED** [review]
Coooooool 🎉
Question: à quoi sert le fichier `newTestStrategy` ?

File: doc/standards/testStrategy/newTestStrategy.md [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+  v1: jest.fn(),
+  v4: jest.fn(mockV4),
+}))
+```
+
+And use `mockV4.mockReturnValueOnce('yourId')` for each uuid needed.
+
+---
+
+## Component test
```
Comment (Lucasbeneston) [inline doc/standards/testStrategy/newTestStrategy.md:153]: Tu dupliques les infos dans 2 fichiers ?

**Review by Lucasbeneston — APPROVED** [review]

File: doc/standards/testStrategy/ComponentTest.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
+# Component test
+
+Do not test component with snapshots
+
+## Navigation Test
+
+Somes examples
```
Comment (voisinhugo) [inline doc/standards/testStrategy/ComponentTest.md:7]:
```suggestion
Some examples
```

File: doc/standards/testStrategy/newTestStrategy.md (hunk 9 above)
Comment (marineshaw) [reply doc/standards/testStrategy/newTestStrategy.md]: Non en effet c'est une erreur (les restes d'avant mon découpage), @anoukhello je supprime le fichier `newTestStrategy`

File: doc/standards/testStrategy/ComponentTest.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  await waitForExpect(() => {
+    expect(mockSignIn).toHaveBeenCalledTimes(1)
+    expect(navigate).ttoHaveBeenCalledTimes(1)
+  })
+})
+```
+
+---
+
+## ToHaveBeenCalled...
```
Comment (voisinhugo) [inline doc/standards/testStrategy/ComponentTest.md:36]: Vu avec Marine, TODO : retrouver la doc là-dessus, si on veut toujours en parler :)

File: doc/standards/testStrategy/ComponentTest.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+---
+
+## Test the responsibility of the components
+
+Ideally, you want to make sure that the tests for a component only test that which is the responsibility of the component. There are several reasons for this:
+
+- With jest, we test our components in a unitary way, so they must be separated from the rest of the code
+- If a component is used in several places and its behavior is tested each time, we have unnecessary duplicates and this creates a waste of time, both in writing these tests and in launching them
+
+How to do it ?
```
Comment (voisinhugo) [inline doc/standards/testStrategy/ComponentTest.md:47]:
```suggestion
How to do it?
```

File: doc/standards/testStrategy/ComponentTest.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  const leftIcon = getByTestId('leftIcon')
+  fireEvent.press(leftIcon)
+
+  expect(goBack).toHaveBeenCalledTimes(1)
+})
+```
+
+```jsx
+it('should redirect to home page when signin is successful', async () => {
+  const { findByText } = renderLogin()
```
Comment (voisinhugo) [inline doc/standards/testStrategy/ComponentTest.md:21]: Est-ce qu'on veut mettre en avant la création de fonction de "render“ comme celle-ci ?
Est-ce que ça serait pas mieux d'avoir ?
```
  const { findByText } = render(<Login />)
```

File: doc/standards/testStrategy/FullPageTest.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
+# Full page test
+
+## Render test
+
+```jsx
+it('should render correctly', async () => {
+  const page = await render(<Page />)
```
Comment (voisinhugo) [inline doc/standards/testStrategy/FullPageTest.md:7]: On veut vraiment ce `await` ? 🤔  (avec son `async`)

File: doc/standards/testStrategy/FullPageTest.md [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+          'aria-allowed-role': { enabled: false },
+        },
+      })
+
+      expect(results).toHaveNoViolations()
+[...]
+```
+
+If you encounter an error with `duplicate-id-aria` this is due to the global uuid mock, we systematically mock `testUuidV4`
+To correct this error it is necessary to mock uuid in your test file like this :
```
Comment (voisinhugo) [inline doc/standards/testStrategy/FullPageTest.md:67]:
```suggestion
To correct this error it is necessary to mock uuid in your test file like this:
```

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
+# General guidelines
+
+## FIRST principles
+
+- Fast : Could be run anytime during the development phase
+- Independant/Isolated : Test should also follow 3A (Arrange, Act, Assert, ou Given-When-Then) and test one functionality
+- Repeatable : Each test sould give the same result each time it runs
+- Self-validating : The output of a test should be "yes" or "no", and not some values to check
+- Thorough : Tests must cover all possibilities and edge cases
+- Timely : Test should be written before their feature (TDD)
```
Comment (voisinhugo) [inline doc/standards/testStrategy/GeneralGuidelines.md:10]: C'est FIRSTT du coup ? 🤔  Dommage ! 🤡

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+---
+
+## When to use .web and .native?
+
+use `MyComponent.native.test.tsx` if
```
Comment (voisinhugo) [inline doc/standards/testStrategy/GeneralGuidelines.md:16]: à la première lecture, j'ai cru que c'était "if … and … and …", alors que c'est plutôt "if … or … or …" :/ 
Peut-être dire "either if" ?
(pareil pour les cas suivants)

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+- a big and complex component
+- a component used in a single page
+
+---
+
+## When to use queryBy, findBy and getBy
+
+- Prove that an element exists with `getBy` / `queryBy` + `expect(…).toBeTruthy()`
+- Prove that an element does not exist with `queryBy` + `expect(…).toBeNull()`
+- Query an element to process some action on it : `getBy` + action
```
Comment (voisinhugo) [inline doc/standards/testStrategy/GeneralGuidelines.md:65]:
```suggestion
- Query an element to process some action on it: `getBy` + action
```

File: doc/standards/testStrategy/GeneralGuidelines.md (hunk 16 above)
Comment (voisinhugo) [reply doc/standards/testStrategy/GeneralGuidelines.md:10]: Sinon, attention, pas d'espace avant ":" en anglais

File: doc/standards/testStrategy/GeneralGuidelines.md (hunk 6 above)
Comment (voisinhugo) [reply doc/standards/testStrategy/GeneralGuidelines.md:66]: J'avoue que je préfère avoir des `findBy` partout, mais je verrai plutôt une reco type "prefer to use `findBy` rather than waitFor(getBy), as it is equivalent". Mais les deux autres me vont aussi

File: doc/standards/testStrategy/GeneralGuidelines.md [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+- a component used in a single page
+
+---
+
+## When to use queryBy, findBy and getBy
+
+- Prove that an element exists with `getBy` + `expect(…).toBeTruthy()`
+- Prove that an element does not exist with `queryBy` + `expect(…).toBeNull()`
+- Query an element to process some action on it with `getBy` + action
+- **Prefer to use** `await findBy`instead of `waitFor(getBy)`or `waitFor(queryBy)`,
```
Comment (anoukhello) [inline doc/standards/testStrategy/GeneralGuidelines.md:66]:
```suggestion
- **Prefer to use** `await findBy`instead of `waitFor(getBy)`or `waitFor(queryBy)` as it is equivalent`,
```
C'est une proposition, moi j'ai découvert assez tard que `findBy` était un raccourci de `waitFor(getBy)`

**Review by anoukhello — APPROVED** [review]
